Missing outputs should be recorded as test errors

bernt-matthias opened the pull request:
from `planemo` the `verify_tool` function is called from [within an try-except block](https://github.com/galaxyproject/planemo/blob/1aa3eb05a97ad20c0be6f6560ab5cec090e76612/planemo/engine/galaxy.py#L109) which silently catches any exception.

Thus any exception raised from within verify_tool will not be detected, i.e. the assertion needs to be moved into `_verify_outputs` (which also seems to make sense by name) in order to make verify_tool record the problem properly.

This only affects tool tests. Guess we do not want to actually forbid tool runs with no outputs.. Even if I don't know what they are used for. 
## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

bernt-matthias committed:
tool verification: move no output assertion

from planemo the verify_tool function is called from
[within an try-except block](https://github.com/galaxyproject/planemo/blob/1aa3eb05a97ad20c0be6f6560ab5cec090e76612/planemo/engine/galaxy.py#L109)
which silently catches any exception.

Thus any exception raised from within verify_tool will not be
detected, i.e. the assertion needs to be moved into
`_verify_outputs` (which also seems to make sense by name)
in order to make verify_tool record the problem properly.

bernt-matthias committed:
adapt output filter test

- remove the output that is always present
- add a test that has no outputs

bernt-matthias committed:
make test a bit more specific

bernt-matthias committed:
register failure to produce output

thereby we implicitly make the assumption that a test needs to produce
an output

also make message more specific

bernt-matthias committed:
register later

mvdbeek commented on lib/galaxy/tool_util/verify/interactor.py:
```diff
@@ -1511,6 +1509,10 @@ def register_exception(e: Exception):
 
+    if not (data_list or data_collection_list):
+        error = AssertionError("Tool produced no output datasets or collections")
+        register_exception(error)
```
```suggestion
    assert data_list or data_collection_list, "Tool produced no output datasets or collections"
```

bernt-matthias replied:
We need to (at least it's the idea of the PR) record the assertion error to check for `expect_test_failure` at the end.

mvdbeek replied:
I don't understand what you're saying. `expect_test_failure` is set by the test definition. What are you trying to fix ?

bernt-matthias replied:
We need to record the exception (i.e. call `register_exception` which stores in `found_exceptions`) in order to check for `expect_test_failure` https://github.com/galaxyproject/galaxy/blob/288a6c989a37be04a357191aeae9711f3b714806/lib/galaxy/tool_util/verify/interactor.py#L1625

If we just raise the Assertion then we have the same behavior as before. If we register it, we can react to this.

mvdbeek commented on test/functional/tools/output_filter.xml:
```diff
@@ -37,134 +36,113 @@ echo 'test' > p2.reverse
             </output>
         </test>
-        <test expect_num_outputs="1">
+        <!-- tool runs with no outputs should fail -->
```
I don't see why that should be the case ?

bernt-matthias replied:
This was not precise enough. Maybe the comment should say: 

```suggestion
        <!-- a test that does not produce any outputs violates the silent assumption that a tool run should have an output, hence expect_test_failure -->
```

Background if planemo finds a test that does not produce outputs it will skip all output checks and just log the assertion to the terminal. So the user will very likely miss this (in particular in CI). 

We might also remove the assertion completely.

Are there tools that do not produce outputs?

bernt-matthias replied:
So .. the tool itself does not fail (I used `expect_test_failure` and not `expect_failure`).

mvdbeek replied:
> a test that does not produce any outputs violates the silent assumption that a tool run should have an output, hence expect_test_failure

what is this silent assumption ? an export tool doesn't need to create outputs

bernt-matthias replied:
Maybe implicit assumption is better. Anyway the export tool is a good example that the current change is not good.

The bottom line is the following. Currently, no check of the outputs is done if a tool does not produce any output but an assertion is raised. This assertion only appears in the logs and (for instance) planemo users don't see it if they are not paying attention to the logs (which probably only a few people do).

In other words, tool tests are successful if the tool test does not produce any output regardless of if there are any test output assumptions.

Or even shorter: the `AssertionError` short circuits all test outputs assumptions. 

... Maybe just remove the assertion?

mvdbeek commented on test/functional/tools/output_filter.xml:
```diff
@@ -1,15 +1,15 @@
+echo '3' > 3 &&
+echo '4' > 4 &&
+echo '5' > 5 &&
+echo 'p1.forward' > p1.forward &&
```
Is that just refactoring or does this prove something in the tests ?

bernt-matthias replied:
Just refactoring. Wanted to be a bit more sure/explicit in the tests about which file is which.

bernt-matthias commented on lib/galaxy/tool_util/verify/interactor.py:
```diff
@@ -1511,6 +1509,10 @@ def register_exception(e: Exception):
+    if not (data_list or data_collection_list):
+        error = AssertionError("Tool produced no output datasets or collections")
+        register_exception(error)
+
```
```suggestion
```

bernt-matthias committed:
move assertion instead of removing it

otherwise the tool test will just return a list out of bounds exception
which is unclear to the user

bernt-matthias commented on lib/galaxy/tool_util/verify/interactor.py:
```diff
@@ -1503,6 +1500,9 @@ def register_exception(e: Exception):
 
+    if not data_list and not data_collection_list:
+        Exception("Job did not produce any outputs")
+
```
```suggestion
```

bernt-matthias committed:
Just use the suggestion

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

mvdbeek commented on lib/galaxy/tool_util/verify/interactor.py:
```diff
@@ -1516,6 +1513,8 @@ def register_exception(e: Exception):
 
     job_stdio = galaxy_interactor.get_job_stdio(job["id"])
 
+    assert data_list or data_collection_list, "Tool produced no output datasets or collections"
```
That's only a bug if we expected an output. Pure export tools don't need to produce outputs.

bernt-matthias replied:
I see.

bernt-matthias committed:
check only presence of outputs mentioned in the tests

bernt-matthias commented on test/functional/tools/output_filter.xml:
```diff
@@ -37,134 +36,113 @@ echo 'test' > p2.reverse
-                    <has_line line="test" />
-                </assert_contents>
-            </output>
+            <assert_stdout>
```
```suggestion
            <output name="out_3">
                <assert_contents>
                    <has_line line="test" />
                </assert_contents>
            </output>
            <assert_stdout>
                <has_n_lines n="0"/>
            </assert_stdout>
```

bernt-matthias committed:
Add back test output

bernt-matthias commented on test/functional/tools/output_filter.xml:
```diff
@@ -37,134 +36,120 @@ echo 'test' > p2.reverse
+                <has_n_lines n="0"/>
+            </assert_stdout>
+                <has_n_lines n="0"/>
+            </assert_stdout>
```
```suggestion
```

bernt-matthias committed:
Update test/functional/tools/output_filter.xml

mvdbeek commented on lib/galaxy/tool_util/verify/interactor.py:
```diff
@@ -1547,23 +1544,29 @@ def register_exception(e: Exception):
+                else:
+                    output_data = data_list[len(data_list) - len(testdef.outputs) + output_index]
+            except IndexError:
+                pass
```
That seems weird, can you add a comment here ? I think that's also why the framework test failed.

bernt-matthias replied:
You are right. How about https://github.com/galaxyproject/galaxy/pull/16094/commits/6ee5b09f95b2fbd06dc59675d72d1dd68dd9431d? Hope this makes it clearer. The comment would have been more or less the content of the exception we are recording here.

bernt-matthias committed:
move registering the exception

then it might be clearer what's going on here

mvdbeek commented on lib/galaxy/tool_util/verify/interactor.py:
```diff
@@ -1541,29 +1538,34 @@ def register_exception(e: Exception):
+                    output_data = data_list[len(data_list) - len(testdef.outputs) + output_index]
+            except IndexError:
+                error = AssertionError(f"Tool did not produce an output with name '{name}' (or at index {output_index})")
+                register_exception(error)
```
`output_data` is still undefined below. maybe set it to None ?

bernt-matthias replied:
This is also in https://github.com/galaxyproject/galaxy/pull/16094/commits/6ee5b09f95b2fbd06dc59675d72d1dd68dd9431d Just before the `try`. Or do I miss something?

mvdbeek replied:
yes, didn't see that.

bernt-matthias committed:
black
